active-rfcs/0000-tables.md

```diff
@@ -0,0 +1,522 @@
+
+In the above approach the text elements with `containerId` as `cell id` will exists in the `json`. This means there will be 4 separate `text` elements in the `json`.
+
+Here is an alternative version - Having `cells` as virtual text elements instead of actual text elements.
+
+```js
+{
+  "id": "table-id",
+  "type": "table",
+  "x": 100,
+  "y": 100,
+  "title": "Table Title",
+  "rows": 2,
+  "columns": 2,
+  "cells": [
+    {
+      "id": "cell-1",
+      "row": 0,
+      "column": 0,
+      "x": 100,
+      "y": 100,
+      "width": 100,
+      "height": 50,
+      "text":"Cell 1",
+      "fontSize": 16;
+      "fontFamily": 1;
+
+    },
+    {
+      "id": "cell-2",
+      "x": 200,
+      "y": 100,
+      "row": 0,
+      "column": 1,
+      "width": 150,
+      "height": 50,
+      "text":"Cell 2",
+      "fontSize": 16;
+      "fontFamily": 1;
+    },
+    {
+      "id": "cell-3",
+      "x": 100,
+      "y": 150,
+      "row": 1,
+      "column": 0,
+      "width": 120,
+      "height": 60,
+      "text":"Cell 3",
+      "fontSize": 16;
+      "fontFamily": 1;
+    },
+    {
+      "id": "cell-4",
+      "x": 220,
+      "y": 150,
+      "row": 1,
+      "column": 1,
+      "width": 130,
+      "height": 60,
+      "text":"Cell 4",
+      "fontSize": 16;
+      "fontFamily": 1;
+    }
+  ],
+  "angle": 0,
+  "strokeColor": "#000000",
+  "backgroundColor": "transparent",
+  "fillStyle": "hachure",
+  "strokeWidth": 1,
+  "strokeStyle": "solid",
+  "roughness": 0,
+  "opacity": 100,
+  "groupIds": [],
+  "seed": 1,
+  "version": 1,
+  "versionNonce": 1,
+  "isDeleted": false,
+  "boundElements": null,
+  "link": null,
+  "locked": false
+}
+```
+
+As you can see above the `cells` will contain all the attributes (only some of text attributes are shown above) and at the time of rendering these text elements will be drawn, however we won't be storing the text elements as separate elements.
+This would simplify the data structure and reduce the hierarchy, thus helping in better maintainance
+
+#### Questions
+
+When `cells` are virtual elements which are just rendered on canvas but physically they don't exist as separate text elements? This will definately simplify the data structure, however there are some unknowns listed below which needs to be evaluated before we move with this approach.
+
+- With above approach of virtual elements, how will impact the overall performance of interacting with tables since we heavily rely on actual elements ?
+- Need to verify if this approach is feasable with current state of TextWYSIWYG
+- How will this impact collaboration ? We may need a custom logic for tables?
+
+Yes we will need write custom logic to handle reconcilation in `tables`, mainly attaching `timestamp` to each cell and writing a custom algorithm to resolve the conflicts and ensuring latest changes are applied to all the collaborators.
+
+### Storing cells with respect to rows to ease reordering of rows and columns
```

Comment:
Hi @Mrazator as per your suggestion, I have added an alternative approach for taking care of swapping rows and columns. 
I have added `index` and `lastUpdated` fields to resolve conflicts during collaboration. 
Let me know what you think

Reply:
I was thinking that in order to support concurrent changes out of the box, we cannot rely on an array at all and thus we would have to replace the whole thing with objects, which gets ugly pretty quickly. Very rough format idea:

```
```js
// initial state
{
  header: {
    "column_A": { value: "A", index: "a0", isDeleted: false },
    "column_B": { value: "B", index: "a1", isDeleted: false },
    "column_C": { value: "C", index": "a2", isDeleted: false },
  },
  rows: {
    "row_0": {
    	cells: {
    	  "cell_0": { value: 1, index: "a0" },
          "cell_1": { value: 2, index: "a1" },
          "cell_2": { value: 3, index: "a2" },
    	},
        index: "a0"
        isDeleted: false,
    },
    "row_1": {
     	cells: {
     	   "cell_0": { value: 4, index: "a0" },
           "cell_1": { value: 5, index: "a1" },
           "cell_2": { value: 6, index: "a2" },
     	},
        index: "a1",
        isDeleted: false,
    },
  }
}
```

That is all based on simple scenarios, which modify columns & rows (notice the likely necessity for soft deletions - `isDeleted`).

A) To the table defined above, "A" added a row `[7, 8, 9]`, "B" removed a column "C", "A" performs undo, "B" performs undo
B) To the table defined above, "A" added a column `D`, "A" performs undo, "B" added a column "E", "A" performs redo

Questionably, similar soft deletions might be / might not be needed on the cell level, but I still didn't go through all the possible cases.

---

Btw. I think we will likely want to avoid relations with created with `boundElements`, as those introduced a lot of challenges during multiplayer undo-redo, so I am guessing it would end in similar cases here. For details check the conflict resolution of relations defined with `boundElements` as part of the `History` tests.

Reply:
@Mrazator in the nested tree structure above, unless the object order is significant, you'd need to sort by `index` first during render if you didn't want to keep explicit `x` & `y` (and potentially `width` & `height`).

At that point we can go with a flat structure and it'll have similar perf profile (except you sort once, but longer array which might be slower depending on number of cells).

Ideally we get rid of sorting.

Below, I'm using a flat structure, but need to track `x`/`y`/`width`/`height`.

```json
{
  "type": "table",
  "cells": {
    "cell-id-1": {
      "id": "cell-id-1",
      "row": 0,
      "column": 0,
      "x": 0,
      "y": 0,
      "width": 100,
      "height": 50
    },
    "cell-id-2": {
      "id": "cell-id-2",
      "row": 0,
      "x": 100,
      "y": 0,
      "column": 1,
      "width": 100,
      "height": 50
    }
  }
}
```

If we want to get rid of `x` & `y`, we'd have to keep the auxiliary `rows` mapping as @Mrazator [suggested](https://github.com/excalidraw/rfcs/pull/1#issuecomment-2228670010)).

As for `width` & `height`, IMO we need to track that regardless, else we'd need to compute it dynamically from the text content and find a `maxHeight` for each row, an `maxWidth` for each column, before we could render any cell.

But by tracking these, a simple change in dimension for once cell would cascade for all the cells in given row/column based on what dimension changed.

If we track `x` & `y` as well, updates would cascade even more.

We would need to sync on property-level instead of element-level.

---

As for `isDeleted`, I'm still not convinced we need it for either.